experimental/InjectiveResolutions/src/MonoidAlgebra.jl

```diff
@@ -0,0 +1,611 @@
+# INPUT:    polyhedral cone C
+# OUTPUT:   set of hyperplanes bounding C
+function get_bounding_hyperplanes(P::Polyhedron)
+  hyperplanes = Vector{HyperplaneQ}()
+  for f in facets(Polyhedron, P)
+    hyperplane = polyhedron(affine_hull(f)[1])
+    A, b = get_hyperplane_H_presentation(hyperplane)
+    push!(hyperplanes, HyperplaneQ(hyperplane, A, b))
+  end
+  return hyperplanes
+end
+
+# TODO: Is this an admissible signature for a method of this function?
+# Should this be moved to the PolyhedralGeometry section?
+# This returns the primitive generator as a `Vector{Int}`.
+function primitive_generator(::Type{Int}, r::AbstractVector{T}) where {T<:RationalUnion}
+  return Vector{Int}(first(primitive_generator_with_scaling_factor(Int, r)))
+end
+
+# This method returns a triple `(v, num, den)` where `v` is a `Vector{Int}` 
+# and `num` and `den` are both `Int`s so that `num//den` is the scaling factor.
+function primitive_generator_with_scaling_factor(
+    ::Type{Int},
+    r::AbstractVector{T}
+  ) where {T<:RationalUnion}
+  @req !is_zero(r) "input must not be a zero vector"
+  first_scaling_factor = lcm(denominator.(r))
+  result = Int[Int(numerator(a)*divexact(first_scaling_factor, denominator(a))) for a in r]
+  g = gcd(result)
+  result = Int[divexact(a, g) for a in result]
+  return Tuple{Vector{Int}, Int, Int}((result, first_scaling_factor, g))
+end
+
+# given a hyperplane, return the H-presentation of it
+# INPUT:    hyperplane h
+# OUTPUT:   matrix A, vector b corresponding to Ax \leq b which defines h
+function get_hyperplane_H_presentation(h::Polyhedron)
+  aff_hull = affine_hull(h).Obj.pm_polytope.AFFINE_HULL
+  _M = Matrix{Rational}(aff_hull)
+  M = hcat(map(row -> reshape(primitive_generator(Int, row), 1, :), eachrow(_M))...)
+  A = [M[:, 2:n_columns(M)]; -M[:, 2:n_columns(M)]]
+  b = [M[:, 1]; -M[:1]]
+  return A, b
+end
+
+# given a polyhedral cone, return the zonotope as in Lemma 3.10 in [HM05]
+# INPUT:    polyhedral cone C  
+# OUTPUT:   zonotope, sum of primitive integer vector ong rays of C 
+function get_zonotope(P::Polyhedron)
+  d = ambient_dim(P)
+  P_rays = [primitive_generator(Int, Vector(r)) for r in rays(P)]
+
+  c = zeros(Int, d)
+  zonotope = convex_hull(zeros(Int, d))
+  for r in P_rays
+    zonotope = zonotope + convex_hull([zeros(Int, 1, d); reshape(r, 1, length(r))])
+    c = c + r
+  end
+  return zonotope, c
+end
+
+@doc raw"""
+    monoid_algebra(B::Matrix{Int},k::Field)
+
+Return the monoid algebra generated by monomials $x^{v_1},\dots,x^{v_n}\in k[x_1,\dots,x_d]$, where $v_1,\dots,v_n\in \mathbb{Z}^d$ are the columns of $M_Q$. 
+
+# Examples
+```jldoctest
+julia> M_Q = [1 0; 0 1]
+2×2 Matrix{Int64}:
+ 1  0
+ 0  1
+
+julia> monoid_algebra(M_Q,QQ)
+monoid algebra over rational field with cone of dimension 2
+```
+"""
+function monoid_algebra(M_Q::Matrix{Int}, k::Field)
+  d = size(M_Q, 1)
+
+  # construct k[t_1,...,t_d]
+  t_vars = [Symbol("t_$i") for i in 1:d]
+  T, t = graded_polynomial_ring(k, t_vars)
```

Comment:
See https://docs.oscar-system.org/stable/DeveloperDocumentation/caching/
```suggestion
  T, t = graded_polynomial_ring(k, t_vars; cached = false)
```